### PR TITLE
ethclient: serialize negative block number as "pending"

### PR DESCRIPTION
**Pull request by sammy007**
https://github.com/ethereum/go-ethereum/issues/21175

### CODE REVIEW

**Comment by sammy007** [comment]
Requires further work, because getting "missing required field 'transactionIndex' for Log" for pending events.

File: core/types/log.go [hunk 1]
```diff
@@ -44,11 +44,11 @@ type Log struct {
 	// hash of the transaction
 	TxHash common.Hash `json:"transactionHash" gencodec:"required"`
 	// index of the transaction in the block
-	TxIndex uint `json:"transactionIndex" gencodec:"required"`
```
Comment (rjl493456442) [inline core/types/log.go:47]: Any particular reason to drop this restriction?

Comment (sammy007) [reply core/types/log.go:47]: No block = no tx index.

File: ethclient/ethclient.go [hunk 2]
```diff
@@ -282,6 +282,9 @@ func toBlockNumArg(number *big.Int) string {
 	if number == nil {
 		return "latest"
 	}
+	if number.Sign() < 0 {
```
Comment (fjl) [inline ethclient/ethclient.go:285]: I think we should be a bit more strict here and only allow -1 to mean pending.
That way, we can use other negative numbers for different things later.

**Pull request merged**